### ISSUE TITLE
#file won't read resources from inside uberjar

### ISSUE THREAD

**Issue by karlmikko** [juxt/aero#12]
using the `#file` reader condition areo fails to read files packaged in an uberjar

**Comment by karlmikko** [comment]
fix idea: add option to tell the reader the file is a resource this would then be passed down to subsequent `#file` readers as the option map is preserved.

``` clojure
(areo/read-config "conf.edn" {:resource true})
```

or pass in the file reader

``` clojure
(areo/read-config "conf.edn" {:file-reader clojure.java.io/resource})
```

**Comment by karlmikko** [comment]
Further - it may be worth making a `#resource` reader condition

**Comment by karlmikko** [comment]
I have been looking at this issue closer over the past week, a few more notes.

`#file` location is always relative to where the binary is called, not where the edn using `#file` is.

`clojure.java.io/resource` returns a `java.net.URL` which the path can be drawn from.

once we have the path - either by extracting it from the URL or directly from a string, we can then store the `parent` directory for use with future calls to `#file` allowing config files to declare paths relative to that  config files location.

I have made a branch with the functionality as described here https://github.com/karlmikko/aero/commit/b98b4557c08ee9b3d71a1355acd9bdd24bb042f5

However this does break current behaviour as where `#file` is relative from is now different.

This does however allow files inside `resources` to refer to each other and not have any trouble when moved to be packaged in a `.jar`. **I have yet to test this inside a jar as it isn't something I know how to do from `lein test`.

The other downside to this branch is currently it will add extra files used for testing to the uberjar as there are files in the resources directory. Also needs the docs to be updated.

**Comment by karlmikko** [comment]
Considering that `#file` is being depreciated in favour of `#include` it is probably a good time to change behaviour of relative paths.

**Comment by tangrammer** [comment]
👍

**Comment by tangrammer** [comment]
Hi Malcolm,
I'm using now the fork of @karlmikko to solve this problem, just wondering about how juxt.pro normally solves this problem and how do you feel with the solution proposed 

thanks a lot!!

**Comment by malcolmsparks** [comment]
I haven't looked at the solution yet but I'm expecting it will be fine to
merge.

As you ask, the juxt.pro website doesn't use uberjars. It's a production
hardened derivative of git-clone-boot-run using systemd and git's
push-to-working-tree feature. Personally I dislike uberjars - they feel too
heavy - but it's a minority opinion!

**Comment by tangrammer** [comment]
sorry, i was meaning `juxt.pro` as a company-client-projects not as a website 😄 

Thanks!

**Comment by karlmikko** [comment]
My solution is actually around maintaining the relative file path, this in turn keeps the path for resources in an uberjar.

**Comment by karlmikko** [comment]
I will need to make another pr though as the code base has moved on quite a bit. Will prob do over the next day or so.

**Comment by malcolmsparks** [comment]
@karlmikko It will be very welcome - I'm completely agree we need to fix this

**Comment by malcolmsparks** [comment]
@tangrammer We have some client projects that use uberjars, and some that use aero, but there is no elements in the set intersection (yet).

**Comment by karlmikko** [comment]
Fixed with #13
